### ISSUE TITLE
Android OOM upon network disconnect

### ISSUE THREAD

**Issue by bubenheimer** [grpc/grpc-java#4860]
With grpc-java 1.15.0 on Android I am seeing out of memory issues when network connectivity goes away. `isReady()` indicates unlimited readiness on the outbound stream for a bidi call until Java runs out of memory.

Here is the relevant code, the complete example is at https://github.com/bubenheimer/grpcsinglestreamingcallflush

    StreamDataGrpc.newStub(channel).streamData(
            new ClientResponseObserver<Item, Item>() {
                private int outCounter = 0;
                private int inCounter = 0;
    
                @Override
                public void beforeStart(
                        final ClientCallStreamObserver<Item> requestStream) {
                    requestStream.setOnReadyHandler(() -> {
                        while (requestStream.isReady()) {
                            final Item item = Item.newBuilder()
                                    .setValue(Integer.toString(outCounter++)).build();
                            requestStream.onNext(item);
                            if (outCounter == 50_000) {
                                requestStream.onCompleted();
                                break;
                            }
                        }
                    });
                }
                @Override
                public void onNext(final Item value) {
                    ++inCounter;
                }
    
                @Override
                public void onError(final Throwable t) {
                    Log.w(TAG, t);
                }
    
                @Override
                public void onCompleted() {
                    Log.i(TAG, "Call completed after " + inCounter + " Items");
                }
            });

where `channel` is more persistently defined as

    channel = AndroidChannelBuilder
                // localhost server from Android emulator
                .forAddress("10.0.2.2", 8082)
                .usePlaintext()
                .context(getApplicationContext())
                .build();

To trigger the issue I do one complete run of the RPC (50,000 messages upstream & 50,000 messages downstream) with the mobile data network on, then I turn mobile data off and do a second run on the same channel. The second run quickly runs out of memory as grpc-java buffers messages it can't send. The channel does not see a break in the connection when mobile data switches off.

I expected that isReady() would take care of this problem, but it did not. I have tried different variations, including one where writing to the stream is done from a separate, dedicated thread, with 1ms delays between each write, to not block anything and avoid races, but the outcome was the same.

**Comment by carl-mastrangelo** [comment]
@bubenheimer Does the same thing happen when using OkHttpChannelBuilder?

**Comment by bubenheimer** [comment]
Yes, the same thing happens with OkHttpChannelBuilder. Depending on device configuration the number of messages may need to be adjusted for OOM.

This is a connection break simulation to manually trigger an undetected broken connection. The connection break remains undetected on the Android emulator, at least for a number of minutes, which is helpful for evaluating more general connection breakage scenarios.

I put the logs on pastebin, might be obvious to a grpc-trained eye what's going on: https://pastebin.com/8SZPmRzm

**Comment by carl-mastrangelo** [comment]
Assigning ejona for triage

**Comment by ejona86** [comment]
Interesting. Sounds like a flow control problem. @bubenheimer, it looks like this never worked for you, as opposed to being a regression. Is that right? (I'm not asking whether it _is_ a regression, but whether you are reporting a regression.)

It failing with plain `OkHttpChannelBuilder` is helpful.

@bubenheimer, what is the HTTP/2 server the client is talking to? That make be a gRPC server, but I am interested in the first "hop" through any proxies and the like (so the proxy a proxy is in use). 

There's a fixed-size 32 KB buffer between the application and network. That's nowhere near large enough to contain all the messages (the accounting includes the 5-byte gRPC header before each message). If the server opens up a 1 MB HTTP/2 window, however, then they would all be queued to be written and the application wouldn't see any flow control. That's why am asking which server is being used. (Due to OkHttp's threading, we may not actually be gaining much by the 32 KB buffer...)

The per-message accounting is probably enough to trigger the OOM. Each message is its own separate Buffer, and not combined with near Buffers while in OkHttpClientStream's `pendingData` (... which probably doesn't matter here) and OutboundFlowController's `pendingWriteQueue` (... which probably also doesn't matter here) and in the Executor queue of `AsyncFrameWriter`. But the big issue is that each Buffer seems is consuming 8 KiB for like ~10 bytes of data.

There's a lot to fix here to reduce memory usage... Unfortunately I think the lack of Buffer-combining in `AsyncFrameWriter` is probably the problem here, and it's probably the most invasive to fix. We should probably move the FrameWriter on the calling thread and instead just have the `Sink` within the Executor. We will need to use a Buffer-aware queue, in order to do combining (no more abusing the queue from the Executor), since the queue itself is where the tons-of-Buffers is sitting. We can't add to the queue in larger batches, as corking is needed for that, and we can't drain the queue quickly because we're blocked on the write.

**Comment by bubenheimer** [comment]
@ejona86 and others: I wrote this test in the last days, so not a known regression.

Just to clarify, the code from the logs uses 100,000 messages instead of the 50,000 from what I posted, and it uses OkHttpChannelBuilder, otherwise it's the same code. The log shows two calls, the first one with mobile data on ("Call finished" on line 22 marks the end), and the second one after I disabled mobile data.

The code runs on the Android emulator (I tried with Android API levels 28 and 23) on my Mac and talks to a simple grpc-java server on the Mac via that 10.0.2.2 IP that identifies the host system to the emulator. The grpc-java server uses Netty and all defaults. The emulator pretends that data goes through a mobile data connection, so it must be doing something special; I don't know the details, or what buffering & proxying may happen in-between.

I also ran this test on a real Android phone (talking to a remote server I set up), and there the broken connection is instantly seen once I turn mobile data off, so it will refuse to send any messages, no OOM.
